californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageIdProvider.java

```diff
@@ -31,25 +37,64 @@
  */
 public class InMemoryMessageIdProvider implements MessageIdProvider {
 
+	enum TrackerMode {
+		NULL, GROUPED, HASHED
+	}
+
 	private final LeastRecentlyUsedCache<InetSocketAddress, MessageIdTracker> trackers;
-	private final NetworkConfig config;
+	private final TrackerMode mode;
+	private final Random random;
+	private final int midGroups;
+	private final long exchangeLifetimeMillis;
 
 	/**
 	 * Creates an new provider for configuration values.
 	 * 
-	 * @param config the configuration to use. In particular, the <em>EXCHANGE_LIFETIME</em>
-	 *         configuration parameter is used as the period of time a message ID is marked as
-	 *         <em>in use</em> when it is allocated for a message exchange.
+	 * The following configuration values are used:
+	 * <ul>
+	 * <li>{@link org.eclipse.californium.core.network.config.NetworkConfig.Keys#MID_PROVIDER_GROUPS}
+	 * - determine the group size for the message IDs. Each group is marked as
+	 * <em>in use</em>, if a MID within the group is used.</li>
+	 * <li>{@link org.eclipse.californium.core.network.config.NetworkConfig.Keys#EXCHANGE_LIFETIME}
+	 * - each group of a message ID returned by <em>getNextMessageId</em> is
+	 * marked as <em>in use</em> for this amount of time (ms).</li>
+	 * <li>{@link org.eclipse.californium.core.network.config.NetworkConfig.Keys#USE_RANDOM_MID_START}
+	 * - if this value is {@code true} then the message IDs returned by
+	 * <em>getNextMessageId</em> will start at a random index. Otherwise the
+	 * first message ID returned will be {@code 0}.</li>
+	 * </ul>
+	 * 
+	 * @param config the configuration to use.
 	 * @throws NullPointerException if the config is {@code null}.
+	 * @throws IllegalArgumentException if the config contains no value tracker mode.
 	 */
 	public InMemoryMessageIdProvider(final NetworkConfig config) {
 		if (config == null) {
 			throw new NullPointerException("Config must not be null");
 		}
-		this.config = config;
-		trackers = new LeastRecentlyUsedCache<>(
-				config.getInt(NetworkConfig.Keys.MAX_ACTIVE_PEERS, 150000),
-				config.getLong(NetworkConfig.Keys.MAX_PEER_INACTIVITY_PERIOD, 10 * 60)); // 10 minutes
+		String textualMode = null;
+		TrackerMode mode = TrackerMode.GROUPED;
+		try {
+			textualMode = config.getString(NetworkConfig.Keys.MID_TRACKER);
+			mode = TrackerMode.valueOf(textualMode);
+		} catch (Throwable t) {
+			if (null == textualMode) {
+				throw new IllegalArgumentException("Tracker mode unknown!");
+			} else {
+				throw new IllegalArgumentException("Tracker mode '" + textualMode + "' not supported!");
+			}
+		}
+		this.mode = mode;
+		this.exchangeLifetimeMillis = config.getLong(NetworkConfig.Keys.EXCHANGE_LIFETIME);
```

Comment:
My idea was mainly triggered by the "mid" intialization, where the code was cloned to the implementations. So currently we use 4 values, two must be evaluated at this level. For the other two, I felt, that the config is somehow more. But it can be changed back for the exchangelifetime and midgroups.